rpmio/rpmpgp.c

```diff
@@ -473,16 +473,15 @@ static int pgpPrtSubType(const uint8_t *h, size_t hlen, pgpSigType sigtype,
 	    for (i = 1; i < plen; i++)
 		pgpPrtVal(" ", pgpKeyServerPrefsTbl, p[i]);
 	    break;
-	case PGPSUBTYPE_SIG_CREATE_TIME:
+	case PGPSUBTYPE_SIG_CREATE_TIME:  /* signature creation time */
+	    if (plen-1 != sizeof(_digp->time))
+		break; /* other lengths not understood */
+	    if (_digp->saved & PGPDIG_SAVED_TIME)
+		return 1; /* duplicate timestamps not allowed */
 	    impl = *p;
-	    if (!(_digp->saved & PGPDIG_SAVED_TIME) &&
-		(sigtype == PGPSIGTYPE_POSITIVE_CERT || sigtype == PGPSIGTYPE_BINARY || sigtype == PGPSIGTYPE_TEXT || sigtype == PGPSIGTYPE_STANDALONE))
```

Comment:
Why is this check no longer needed?

Reply:
This check violated RFC 4880 §5.2.3.4, which applies to all types of signatures.

Reply:
IMHO the point of the check is not to make sure that a creation time is present, but to ignore signatures that rpm does not care about.

Reply:
So you care about positive certifications, binary and text signatures, and standalone signatures?  That list seems dubious:

- First, there are more certifications that can be used by implementations as binding (self) signatures.  Note the "most" in Section [5.2.1 of RFC4880](https://datatracker.ietf.org/doc/html/rfc4880#section-5.2.1):

>       Most OpenPGP implementations make their "key signatures" as 0x10
>       certifications.  Some implementations can issue 0x11-0x13
>       certifications, but few differentiate between the types.

- I doubt that you care about standalone signatures.  Those are like binary signatures over a zero-length documents.  If RPM cares, I'd be curious what the use case is.
- You should really care about subkey binding signatures.  How else would you know that a subkey is eligible to make signatures in the name of the certificate, i.e. whether 
  -  the subkey is bound to a certificate at the time the signature in question is created, and
  -  the subkey is marked as signing-capable?

This change looks good.

Reply:
Oh, the list is certainly dubious.

"Care about" here means that the creation time is used calculate the "release" element of the synthetic pubkey rpm package. There's nothing else rpm does with the creation time of pubkey signatures.

Reply:
> * First, there are more certifications that can be used by implementations as binding (self) signatures.  Note the "most" in Section [5.2.1 of RFC4880](https://datatracker.ietf.org/doc/html/rfc4880#section-5.2.1):
> 
> > ```
> >   Most OpenPGP implementations make their "key signatures" as 0x10
> >   certifications.  Some implementations can issue 0x11-0x13
> >   certifications, but few differentiate between the types.
> > ```

RPM does not implement self-signatures at all.  RPM basically implements a subset of `gpgv`.

> * You should really care about subkey binding signatures.  How else would you know that a subkey is eligible to make signatures in the name of the certificate, i.e. whether
>   
>   * the subkey is bound to a certificate at the time the signature in question is created, and

How should RPM determine this?  Is it by checking that the signature creation time was after the binding signature creation time?  In the presence of multiple binding signatures, is it okay to only use the first one?

>   * the subkey is marked as signing-capable?

RPM does not know either of these.  I actually consider this to be a security vulnerability, but I do not know if upstream does.  In any case there is no point in an embargo since this is already public.

> "Care about" here means that the creation time is used calculate the "release" element of the synthetic pubkey rpm package. There's nothing else rpm does with the creation time of pubkey signatures.

In this case RPM should use a completely separate field for this.  The actual time is important for the reasons @teythoon mentioned.

Reply:
> > * First, there are more certifications that can be used by implementations as binding (self) signatures.  Note the "most" in Section [5.2.1 of RFC4880](https://datatracker.ietf.org/doc/html/rfc4880#section-5.2.1):
> > 
> > > ```
> > >   Most OpenPGP implementations make their "key signatures" as 0x10
> > >   certifications.  Some implementations can issue 0x11-0x13
> > >   certifications, but few differentiate between the types.
> > > ```
> 
> RPM does not implement self-signatures at all. RPM basically implements a subset of `gpgv`.

`gpgv` canonicalizes certificates.  Not doing so is unsafe.  Saying RPM implements a subset of that is like saying `/bin/true` implements a subset of `gpgv`.

> > * You should really care about subkey binding signatures.  How else would you know that a subkey is eligible to make signatures in the name of the certificate, i.e. whether
> >   
> >   * the subkey is bound to a certificate at the time the signature in question is created, and
> 
> How should RPM determine this? Is it by checking that the signature creation time was after the binding signature creation time? In the presence of multiple binding signatures, is it okay to only use the first one?

No.  You need to consider the newest valid signature (i.e. non-expired, non-revoked, policy likes it, maths checks out) with a creation time prior to the target signature's creation time.

> > * the subkey is marked as signing-capable?
> 
> RPM does not know either of these. I actually consider this to be a security vulnerability, but I do not know if upstream does. In any case there is no point in an embargo since this is already public.

It is a security vulnerability, but upstream disagrees.

(But, contemporary SLES likes MD5 signatures in FIPS mode, so the bar isn't high.)

> > "Care about" here means that the creation time is used calculate the "release" element of the synthetic pubkey rpm package. There's nothing else rpm does with the creation time of pubkey signatures.
> 
> In this case RPM should use a completely separate field for this. The actual time is important for the reasons @teythoon mentioned.

See, this is the problem.  You point out a problem with their OpenPGP implementation, and get told why it isn't a problem for RPM, or why it is okay in this case, or why you just haven't looked at it from their perspective enough.

I have cared for the most widely used OpenPGP implementation.  Then, I wrote a new implementation from scratch that the creator of the protocol calls the most advanced implementation out there.  I have written the most comprehensive OpenPGP test suite ever created that found countless bugs across multiple implementations.

Take my advice or leave it.  I don't have time to argue why RPM is so special that it doesn't need to do what everyone else absolutely needs to do.

Reply:
@teythoon _All_ of your criticisms are valid.  Fixing RPM’s implementation is a _massive_ amount of work, though.  So I am trying to do the best I can in my spare time.

Right now, for example, I am not even sure where RPM should _store_ revocation information.  In the rpmdb?  In the filesystem somewhere?

> > > * the subkey is marked as signing-capable?
> > 
> > RPM does not know either of these. I actually consider this to be a security vulnerability, but I do not know if upstream does. In any case there is no point in an embargo since this is already public.
> 
> It is a security vulnerability, but upstream disagrees.

Have you made a report to <secalert@redhat.com>?

> I have written the most comprehensive OpenPGP test suite ever created that found countless bugs across multiple implementations.

Is this test suite open source?  I would like to use it to test rpm-oxide’s signature parsing code.

> `gpgv` canonicalizes certificates. Not doing so is unsafe. Saying RPM implements a subset of that is like saying `/bin/true` implements a subset of `gpgv`.

Is there a specification for certificate canonicalization anywhere?

Reply:
> @teythoon _All_ of your criticisms are valid. Fixing RPM’s implementation is a _massive_ amount of work, though. So I am trying to do the best I can in my spare time.
> 
> Right now, for example, I am not even sure where RPM should _store_ revocation information. In the rpmdb? In the filesystem somewhere?

Revocations are signatures that can be stored with the certificate itself.
 
> > > > * the subkey is marked as signing-capable?
> > > 
> > > 
> > > RPM does not know either of these. I actually consider this to be a security vulnerability, but I do not know if upstream does. In any case there is no point in an embargo since this is already public.
> > 
> > 
> > It is a security vulnerability, but upstream disagrees.
> 
> Have you made a report to [secalert@redhat.com](mailto:secalert@redhat.com)?

No.  Feel free to.

> > I have written the most comprehensive OpenPGP test suite ever created that found countless bugs across multiple implementations.
> 
> Is this test suite open source?

Of course: 

- https://gitlab.com/sequoia-pgp/openpgp-interoperability-test-suite
- https://tests.sequoia-pgp.org/

> I would like to use it to test rpm-oxide’s signature parsing code.

Great!  The easiest way is to implement https://docs.rs/sop . Doing so will give you a SOP cli interface for free that we can plug into the test suite.

> > `gpgv` canonicalizes certificates. Not doing so is unsafe. Saying RPM implements a subset of that is like saying `/bin/true` implements a subset of `gpgv`.
> 
> Is there a specification for certificate canonicalization anywhere?

Not that I'm aware, but Paul has a good write-up on how to properly verify signatures: https://blog.jabberhead.tk/2021/04/03/why-signature-verification-in-openpgp-is-hard/

In terms of looking at code, I recommend looking at Sequoia and PGPainless.

Reply:
> > > > `gpgv` canonicalizes certificates. Not doing so is unsafe. Saying RPM implements a subset of that is like saying `/bin/true` implements a subset of `gpgv`.
> > > 
> > > 
> > > Is there a specification for certificate canonicalization anywhere?
> > 
> > 
> > Not that I'm aware, but Paul has a good write-up on how to properly verify signatures: https://blog.jabberhead.tk/2021/04/03/why-signature-verification-in-openpgp-is-hard/
> 
> Would it be possible for the specification to include detailed algorithms for verifying a signature, canonicalizing a certificate, determining when a signature is malformed and should be rejected, and other important facets of OpenPGP?

Historically, the OpenPGP spec has mostly concerned itself with syntax, not semantics.  Therefore, implementing OpenPGP securely requires a lot of tribal knowledge.  Reading RFC4880 and implementing something based only on that reading is a recipe for disaster.

Many things are possible.  We have often talked about writing up that tribal knowledge in a separate document, like an annotated version of the RFC.  That takes people with that kind of knowledge a lot of time.  As always, someone needs to do it.

I have seen many horrible half-baked implementations, you know, like the one in RPM.  Sure, you can spend a lot of time improving that, but in the end, you shouldn't even attempt to implement a complex security-relevant protocol, you should use a library.  You don't implement TLS, you don't implement XZ, RSA, EdDSA.  Why would you even think that it is okay to implement OpenPGP?

> > > Is this test suite open source?
> > 
> > 
> > Of course:
> > 
> > * https://gitlab.com/sequoia-pgp/openpgp-interoperability-test-suite
> > * https://tests.sequoia-pgp.org/
> 
> Thanks! Last I checked there are a few cases where improvements could be made. You may consider these feature requests, or I can file an issue on GitLab for them.
> 
>     * Some implementations, by design, only implement a subset of OpenPGP.  For instance, a package manager only needs to implement signature verification, so there is no reason to implement other parts of OpenPGP such as signing, encryption, or decryption.  It would be nice to split the test suite in such a way as to be able to easily exclude tests that simply are not applicable for a given implementation.

Yeah, that is half-way implemented.  It "only" requires annotating each test with the required operations, then selecting to run just those tests.  But, I'm very skeptical about these point solution implementations.  RPM's is the first verification-only implementation I have encountered.  And, I have plugged it into the test suite and it was informative even though some tests were not applicable.

>     * It would be nice to have more tests for very malformed packets and subpackets, such as ones that might be produced by a fuzzer.  This is very important for finding and fixing memory unsafety and other bugs in implementations.  In RPM-Oxide, for instance, I am much less worried about wrongly rejecting a good signature than about wrongly accepting a bad one.

The test suite is about semantics of OpenPGP, not about the memory safety of implementations.  I think implementations are better off with compiler-supported fuzzing techniques.  Having said that, Sequoia includes a protocol-aware fuzzer and uses it in its tests.  It could conceivably be adapted to test another implementation.

Reply:
> You don't implement TLS, you don't implement XZ, RSA, EdDSA. Why would you even think that it is okay to implement OpenPGP?

Rpm implements its own because there hasn't been any library it could use, language, licenses and all considered. Back in the day, rpm called out to gpg to verify, but when it started verifying signatures on rpmdb queries that became a performance bottleneck. I presume, because this was all well before my time.

Reply:
In any case, this discussion needs to go somewhere else.